### PR TITLE
fix(cli): Linting improvements

### PR DESCRIPTION
**Pull request by arckey**
Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

* switched to use `go-color`
* print linting results after each source (file), instead of waiting for all linting to finish.
* changed `--format` to `-o/--output`
* `argo lint` now lints all manifest kinds by default and can be configured to lint specific kinds with: `--lint-kinds`

### CODE REVIEW

File: cmd/argo/commands/clustertemplate/lint.go
Comment (simster7) [inline cmd/argo/commands/clustertemplate/lint.go:49]: Why this change? Is this what we do in Argo CD or something?

File: cmd/argo/commands/lint.go
Comment (simster7) [inline cmd/argo/commands/lint.go:34]: nice

File: cmd/argo/lint/formatter_pretty.go
Comment (simster7) [inline cmd/argo/lint/formatter_pretty.go:17]: Is there no `color.Ize(color.Clear)`?

File: docs/cli/argo.md
Comment (simster7) [inline docs/cli/argo.md:110]: I'm thinking if we should still even support `argo {template,cron,cluster-template} lint` if `argo lint` can now lint everything? Maybe "soft link" those commands to `argo lint` for compatibility?

File: cmd/argo/commands/lint.go
Comment (alexec) [inline cmd/argo/commands/lint.go:49]: just kinds would be ok?

File: cmd/argo/commands/clustertemplate/lint.go
Comment (arckey) [reply cmd/argo/commands/clustertemplate/lint.go:49]: @alexec suggested that we should switch from `format` to `output` in the [last PR](https://github.com/argoproj/argo-workflows/pull/5095#issuecomment-782241555), but I didn't have time to change it before it was merged.

File: cmd/argo/commands/lint.go
Comment (arckey) [reply cmd/argo/commands/lint.go]: done

File: cmd/argo/lint/formatter_pretty.go
Comment (arckey) [reply cmd/argo/lint/formatter_pretty.go]: fixed!

File: docs/cli/argo.md
Comment (arckey) [reply docs/cli/argo.md:110]: I thought making the other commands just call [this function](https://github.com/argoproj/argo-workflows/pull/5224/files#diff-c9b1f50c36281159db1f24f20050bb13512a9cea0749bdeb975c0ff8688712eaR76) with their relative kinds, but that would require me to move this function to `github.com/argoproj/argo-workflows/v3/cmd/argo/lint` because if I leave it where it is now it would cause a circular dependency.
Do you think I should do it?

File: cmd/argo/commands/clustertemplate/lint.go
Comment (simster7) [reply cmd/argo/commands/clustertemplate/lint.go:49]: SGTM

File: docs/cli/argo.md
Comment (simster7) [reply docs/cli/argo.md:110]: I think that would be good. @alexec thoughts?

**Review by simster7 — APPROVED** [review]
LGTM. Letting @alexec also give a final review before merge

File: cmd/argo/commands/lint.go
Comment (alexec) [inline cmd/argo/commands/lint.go:62]: these look wrong to me, I don't think they hyphens in them?
also, I think it it `wftmpl`? check the CRD YAML please

File: cmd/argo/lint/lint.go
Comment (alexec) [inline cmd/argo/lint/lint.go:39]: I think this probably be a parameter to `Summarzie`?

File: cmd/argo/lint/lint_test.go
Comment (alexec) [inline cmd/argo/lint/lint_test.go:71]: minor - but you could generate your mock using `mockery` see `ContainerRuntimeExecutor` for example

File: docs/cli/argo.md
Comment (alexec) [reply docs/cli/argo.md:110]: I agree-I think we should try and use the same code everywhere

**Review by alexec — CHANGES_REQUESTED** [review]

File: docs/cli/argo.md
Comment (arckey) [reply docs/cli/argo.md:110]: done

File: cmd/argo/lint/lint.go
Comment (arckey) [reply cmd/argo/lint/lint.go]: Since we now call [RunLint](https://github.com/argoproj/argo-workflows/pull/5224/files#diff-e0caade70a58e8cd2f450d1f3e1baa498d2ac0e5713321d8169f0a747ee93f90R91) we no longer have to expose [Lint](https://github.com/argoproj/argo-workflows/pull/5224/files#diff-e0caade70a58e8cd2f450d1f3e1baa498d2ac0e5713321d8169f0a747ee93f90R115) and `LintOptions`. However, I think it would be useful to expose `Lint` from this package, because it could be utilized by 3rd party tools that use argo as a go module. WDYT?

File: cmd/argo/lint/lint.go
Comment (alexec) [inline cmd/argo/lint/lint.go:319]: Plural?

File: cmd/argo/lint/lint.go
Comment (alexec) [inline cmd/argo/lint/lint.go:248]: !!

File: cmd/argo/lint/lint.go
Comment (alexec) [inline cmd/argo/lint/lint.go:119]: I’d probably make these both mandatory, but provide null defaults.

File: cmd/argo/lint/lint.go
Comment (alexec) [inline cmd/argo/lint/lint.go:93]: I think if this is to work in third party software, it needs to return err rather than panic.

Comment (arckey) [reply cmd/argo/lint/lint.go:93]: Actually I meant to keep `RunLint` as a function to be called our CLI, and `Lint` can be called by other libraries to analyze the lint results and do whatever they want with them.

File: cmd/argo/lint/lint.go
Comment (arckey) [reply cmd/argo/lint/lint.go:119]: that's what I have now, no?

File: cmd/argo/lint/lint.go
Comment (arckey) [reply cmd/argo/lint/lint.go]: Makes sense, done!

**Comment by alexec** [comment]
bump! please dismiss "requested changes" to get a new review

File: cmd/argo/commands/lint.go
Comment (alexec) [inline cmd/argo/commands/lint.go:30]: nice example

**Review by alexec — APPROVED** [review]

**Pull request merged**
